sw/device/lib/dif/dif_resetmgr.c

```diff
@@ -0,0 +1,112 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_resetmgr.h"
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/mmio.h"
+#include "rstmgr_regs.h"  // Generated.
+
+// Register fields that are used in multiple functions.
+
+static const bitfield_field32_t resetmgr_causes_field = {
+  .mask = RSTMGR_RESET_INFO_CAUSE_MASK,
+  .index = RSTMGR_RESET_INFO_CAUSE_OFFSET,
+};
+
+/**
+ * Reset Manager public reset cause enum map to the DIF internal definitions.
+ *
+ * These definitions represent the possible reset causes in the info register
+ * cause field.
+ *
+ * Note: reset causes are not mutually exclusive, so are implemented as a
+ *       bitfield.
+ */
+static const uint32_t resetmgr_causes_map[] = {
+  [kDifResetmgrCausePowerUp] = RSTMGR_RESET_INFO_CAUSE_POR,
+  [kDifResetmgrCauseLowPowerExit] = RSTMGR_RESET_INFO_CAUSE_LOW_POWER_EXIT,
+  [kDifResetmgrCauseWatchdog] = RSTMGR_RESET_INFO_CAUSE_WATCHDOG,
+  [kDifResetmgrCauseSecurityEscalation] = RSTMGR_RESET_INFO_CAUSE_ESCALATION,
+  [kDifResetmgrCauseNonDebugModuleRequest] =
+  RSTMGR_RESET_INFO_CAUSE_NON_DEBUG_MODULE,
+};
```

Comment:
This and the `dif_resetmgr_causes_get` function get much simpler if you move this to a packed bitfield. Having now read the hjson for the hardware this always fits in five bits, which inclines me towards bitfields even more here.

Reply:
Will change to bitfields in the next push, as per previous comment.

Reply:
> This and the dif_resetmgr_causes_get function get much simpler if you move this to a packed bitfield.

Not sure, as we still have to map public enum to private register field offsets. We cannot use the register definitions in the `.h` file. We could hardcode the bit indexes, as:

```
typedef enum dif_resetmgr_cause {
  kDifResetmgrCausePowerUp = 1,
  kDifResetmgrCauseLowPowerExit = 2,
  kDifResetmgrCauseWatchdog = 4,
  kDifResetmgrCauseSecurityEscalation = 8,
  kDifResetmgrCauseNonDebugModuleRequest = 16,
  kDifResetmgrCauseLast = kDifResetmgrCauseNonDebugModuleRequest,
} dif_resetmgr_cause_t;
```
to be the same as:
```
#define RSTMGR_RESET_INFO_CAUSE_POR 0x1
#define RSTMGR_RESET_INFO_CAUSE_LOW_POWER_EXIT 0x2
#define RSTMGR_RESET_INFO_CAUSE_WATCHDOG 0x4
#define RSTMGR_RESET_INFO_CAUSE_ESCALATION 0x8
#define RSTMGR_RESET_INFO_CAUSE_NON_DEBUG_MODULE 0x10
```

But to me that doesn't seem all that great.

Reply:
I think there's no requirement to have the values be the same. If we have a function that does:

```c
if (hw_cause & RSTMGR_RESET_INFO_CAUSE_POR) {
  cause |= kDifResetmgrCausePowerUp;
}
```

Then I think that's a perfectly reasonable conversion.

Reply:
Okay, sounds good.

Reply:
@lenary just suggested to make the values match the autogenerated definitions in the source file with static_assert. I think that is probably a better solution, so I will do that (avoids unnecessary runtime matching).